zonec.c

```diff
@@ -753,6 +754,389 @@ zparser_conv_nsec(region_type *region,
+		zc_error_prev_line("invalid base64 data");
+	} else {
+		r = alloc_rdata(region, 2 * sizeof(uint16_t) + wire_len);
+		r[1] = htons(SVCB_KEY_ECHCONFIG);
+		r[2] = htons(wire_len);
+		memcpy(&r[3], buffer, wire_len);
+	}
+
+	return r;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_alpn_value(region_type *region,
+		const char *val, size_t val_len)
+{
+	uint16_t *r;
+	char *str_dst;
+	const char *next_str;
+	size_t str_len;
+
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + val_len + 1);
+	r[1] = htons(SVCB_KEY_ALPN);
+	r[2] = htons(val_len + 1);
+	str_dst = (void *)&r[3];
+
+	while (val_len) {
+		str_len = (next_str = strchr(val, ','))
+		        ? (size_t)(next_str - val) : val_len;
+
+		/* @TODO: deal with escaped comma's */
+		if (str_len > 255) {
+			zc_error_prev_line("alpn strings need to be"
+					   " smaller than 255 chars");
+			return r;
+		}
+		*str_dst++ = str_len;
+		memcpy(str_dst, val, str_len);
+		str_dst += str_len;
+
+		if (!next_str)
+			break;
+
+		val_len -= next_str - val + 1;
+		val = next_str + 1; /* skip the comma */
+	}
+	return r;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_key_value(region_type *region,
+    const char *key, size_t key_len, const char *val, size_t val_len)
+{
+	uint16_t svcparamkey = svcbparam_lookup_key(key, key_len);
+	uint16_t *r;
+	
+	switch (svcparamkey) {
+	case SVCB_KEY_PORT:
+		return zparser_conv_svcbparam_port_value(region, val);
+	case SVCB_KEY_IPV4HINT:
+		return zparser_conv_svcbparam_ipv4hint_value(region, val);
+	case SVCB_KEY_IPV6HINT:
+		return zparser_conv_svcbparam_ipv6hint_value(region, val);
+	case SVCB_KEY_MANDATORY:
+		return zparser_conv_svcbparam_mandatory_value(region, val, val_len);
+	case SVCB_KEY_NO_DEFAULT_ALPN:
+		zc_error_prev_line("no-default-alpn should not have a value\n");
+		break;
+	case SVCB_KEY_ECHCONFIG:
+		return zparser_conv_svcbparam_echconfig_value(region, val);
+	case SVCB_KEY_ALPN:
+		return zparser_conv_svcbparam_alpn_value(region, val, val_len);
+	default:
+		break;
+	}
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + val_len);
+	r[1] = htons(svcparamkey);
+	r[2] = htons(val_len);
+	memcpy(r + 3, val, val_len);
+	return r;
+}
+
+uint16_t *
+zparser_conv_svcbparam(region_type *region, const char *key, size_t key_len
+                                          , const char *val, size_t val_len)
+{
+	const char *eq;
+	uint16_t *r;
+
+	/* Form <key>="<value>" (or at least with quoted value) */
+	if (val) {
+		/* Does key end with '=' */
+		if (key_len && key[key_len - 1] == '=')
+			return zparser_conv_svcbparam_key_value(
+			    region, key, key_len - 1, val, val_len);
+
+		zc_error_prev_line( "SvcParam syntax error in param: %s\"%s\""
+		                  , key, val);
+	}
+	/* assert(val == NULL); */
+	if ((eq = memchr(key, '=', key_len))) {
+		size_t new_key_len = eq - key;
+
+		return zparser_conv_svcbparam_key_value(
+		    region, key, new_key_len, eq+1, key_len - new_key_len - 1);
+	}
+	/* SvcParam is only a SvcParamKey */
```

Comment:
So this looks like indeed a echconfig (and other SvcParams that have values) without value is possible... maybe this is just okay... WDYT @TCY16 ?

Reply:
resolved in latest commit